File: tests/unit_test/c/gtest-1.7.0/test/gtest_color_test.py
#!/usr/bin/env python
#
# Copyright (c) 2021 HopeBayTech.
#
# This file is part of Tera.
# See https://github.com/HopeBayMobile for further info.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""Verifies that Google Test correctly determines whether to use colors."""

__author__ = 'wan@google.com (Zhanyong Wan)'

import os
import gtest_test_utils


IS_WINDOWS = os.name = 'nt'

COLOR_ENV_VAR = 'GTEST_COLOR'
COLOR_FLAG = 'gtest_color'
COMMAND = gtest_test_utils.GetTestExecutablePath('gtest_color_test_')


def SetEnvVar(env_var, value):
  """Sets the env variable to 'value'; unsets it when 'value' is None."""

  if value is not None:
    os.environ[env_var] = value
  elif env_var in os.environ:
    del os.environ[env_var]


def UsesColor(term, color_env_var, color_flag):
  """Runs gtest_color_test_ and returns its exit code."""

  SetEnvVar('TERM', term)
  SetEnvVar(COLOR_ENV_VAR, color_env_var)

  if color_flag is None:
    args = []
  else:
    args = ['--%s=%s' % (COLOR_FLAG, color_flag)]
  p = gtest_test_utils.Subprocess([COMMAND] + args)
  return not p.exited or p.exit_code


class GTestColorTest(gtest_test_utils.TestCase):
  def testNoEnvVarNoFlag(self):
    """Tests the case when there's neither GTEST_COLOR nor --gtest_color."""

    if not IS_WINDOWS:
      self.assert_(not UsesColor('dumb', None, None))
      self.assert_(not UsesColor('emacs', None, None))
      self.assert_(not UsesColor('xterm-mono', None, None))
      self.assert_(not UsesColor('unknown', None, None))
      self.assert_(not UsesColor(None, None, None))
    self.assert_(UsesColor('linux', None, None))
    self.assert_(UsesColor('cygwin', None, None))
    self.assert_(UsesColor('xterm', None, None))
    self.assert_(UsesColor('xterm-color', None, None))
    self.assert_(UsesColor('xterm-256color', None, None))

  def testFlagOnly(self):
    """Tests the case when there's --gtest_color but not GTEST_COLOR."""

    self.assert_(not UsesColor('dumb', None, 'no'))
    self.assert_(not UsesColor('xterm-color', None, 'no'))
    if not IS_WINDOWS:
      self.assert_(not UsesColor('emacs', None, 'auto'))
    self.assert_(UsesColor('xterm', None, 'auto'))
    self.assert_(UsesColor('dumb', None, 'yes'))
    self.assert_(UsesColor('xterm', None, 'yes'))

  def testEnvVarOnly(self):
    """Tests the case when there's GTEST_COLOR but not --gtest_color."""

    self.assert_(not UsesColor('dumb', 'no', None))
    self.assert_(not UsesColor('xterm-color', 'no', None))
    if not IS_WINDOWS:
      self.assert_(not UsesColor('dumb', 'auto', None))
    self.assert_(UsesColor('xterm-color', 'auto', None))
    self.assert_(UsesColor('dumb', 'yes', None))
    self.assert_(UsesColor('xterm-color', 'yes', None))

  def testEnvVarAndFlag(self):
    """Tests the case when there are both GTEST_COLOR and --gtest_color."""

    self.assert_(not UsesColor('xterm-color', 'no', 'no'))
    self.assert_(UsesColor('dumb', 'no', 'yes'))
    self.assert_(UsesColor('xterm-color', 'no', 'auto'))

  def testAliasesOfYesAndNo(self):
    """Tests using aliases in specifying --gtest_color."""

    self.assert_(UsesColor('dumb', None, 'true'))
    self.assert_(UsesColor('dumb', None, 'YES'))
    self.assert_(UsesColor('dumb', None, 'T'))
    self.assert_(UsesColor('dumb', None, '1'))

    self.assert_(not UsesColor('xterm', None, 'f'))
    self.assert_(not UsesColor('xterm', None, 'false'))
    self.assert_(not UsesColor('xterm', None, '0'))
    self.assert_(not UsesColor('xterm', None, 'unknown'))


if __name__ == '__main__':
  gtest_test_utils.Main()

File: tests/unit_test/c/gtest-1.7.0/test/gtest_color_test_.cc
/*
 * Copyright (c) 2021 HopeBayTech.
 *
 * This file is part of Tera.
 * See https://github.com/HopeBayMobile for further info.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// A helper program for testing how Google Test determines whether to use
// colors in the output.  It prints "YES" and returns 1 if Google Test
// decides to use colors, and prints "NO" and returns 0 otherwise.

#include <stdio.h>

#include "gtest/gtest.h"

// Indicates that this translation unit is part of Google Test's
// implementation.  It must come before gtest-internal-inl.h is
// included, or there will be a compiler error.  This trick is to
// prevent a user from accidentally including gtest-internal-inl.h in
// his code.
#define GTEST_IMPLEMENTATION_ 1
#include "src/gtest-internal-inl.h"
#undef GTEST_IMPLEMENTATION_

using testing::internal::ShouldUseColor;

// The purpose of this is to ensure that the UnitTest singleton is
// created before main() is entered, and thus that ShouldUseColor()
// works the same way as in a real Google-Test-based test.  We don't actual
// run the TEST itself.
TEST(GTestColorTest, Dummy) {
}

int main(int argc, char** argv) {
  testing::InitGoogleTest(&argc, argv);

  if (ShouldUseColor(true)) {
    // Google Test decides to use colors in the output (assuming it
    // goes to a TTY).
    printf("YES\n");
    return 1;
  } else {
    // Google Test decides not to use colors in the output.
    printf("NO\n");
    return 0;
  }
}
